### PR TITLE
add wasm filter piugin

### PR DESCRIPTION
**Pull request by jiuxia211**
### Which issue(s) this PR fixes:

Fixes #

### Does this PR introduced a user-facing change?

The wasm file needs to be mounted to the fluent bit. I tried writing the wasm file to secret or configmap, but it was too long, and I ended up mimicking the way Position DB was mounted.
```release-note

```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by benjaminhuo** [comment]
@wanjunlei @wenchajun pls help to review

File: apis/fluentbit/v1alpha2/plugins/filter/wasm.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	kvs := params.NewKVs()
+	err := w.AddCommonParams(kvs)
+	if err != nil {
+		return kvs, err
+	}
+	if w.WasmPath != "" {
+		kvs.Insert("Wasm_Path", w.WasmPath)
+	}
+	if w.EventFormat != "" {
+		kvs.Insert("EventFormat", w.EventFormat)
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/plugins/filter/wasm.go:46]: `EventFormat` => `Event_Format`?

File: apis/fluentbit/v1alpha2/plugins/filter/wasm.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+type Wasm struct {
+	plugins.CommonParams `json:",inline"`
+	// Path to the built Wasm program that will be used. This can be a relative path against the main configuration file.
+	WasmPath string `json:"wasmPath,omitempty"`
+	// Define event format to interact with Wasm programs: msgpack or json. Default: json
+	EventFormat string `json:"eventFormat,omitempty"`
+	// Wasm function name that will be triggered to do filtering. It's assumed that the function is built inside the Wasm program specified above.
+	FunctionName string `json:"functionName,omitempty"`
+	// Specify the whitelist of paths to be able to access paths from WASM programs.
+	AccessiblePaths string `json:"accessiblePaths,omitempty"`
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/plugins/filter/wasm.go:21]: Changing AccessiblePaths to an array is more user friendly.

File: apis/fluentbit/v1alpha2/fluentbit_types.go [hunk 3]
```diff
@@ -107,6 +107,8 @@ type FluentBitSpec struct {
 	SchedulerName string `json:"schedulerName,omitempty"`
 	// Optional duration in seconds the pod needs to terminate gracefully. Value must be non-negative integer.
 	TerminationGracePeriodSeconds *int64 `json:"terminationGracePeriodSeconds,omitempty"`
+	// Storage for wasm file. You will use it if wasm filter is enabled.
+	WasmFile corev1.VolumeSource `json:"wasmFile,omitempty"`
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/fluentbit_types.go:111]: [Volumes  and VolumesMounts](https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/fluentbit_types.go#L70-L72) can be used to mount wasm files. Does it need to add additional fields?

**Review by wanjunlei — CHANGES_REQUESTED** [review]

File: apis/fluentbit/v1alpha2/fluentbit_types.go (hunk 3 above)
Comment (jiuxia211) [reply apis/fluentbit/v1alpha2/fluentbit_types.go:111]: I think it doesn't need additional fields, just specify the hostPath of the wasm file. 
Like this:
```
wasmFile:
    hostPath:
      path: /etc/wasm/  
```

File: pkg/operator/daemonset.go [hunk 4]
```diff
@@ -143,6 +143,17 @@ func MakeDaemonSet(fb fluentbitv1alpha2.FluentBit, logPath string) *appsv1.Daemo
...
 
+	// Mount Wasm Filter
+	if fb.Spec.WasmFile != (corev1.VolumeSource{}) {
+		ds.Spec.Template.Spec.Volumes = append(ds.Spec.Template.Spec.Volumes, corev1.Volume{
+			Name:         "wasm",
+			VolumeSource: fb.Spec.WasmFile,
+		})
+		ds.Spec.Template.Spec.Containers[0].VolumeMounts = append(ds.Spec.Template.Spec.Containers[0].VolumeMounts, corev1.VolumeMount{
+			Name:      "wasm",
+			MountPath: "/fluent-bit/wasm",
```
Comment (wanjunlei) [inline pkg/operator/daemonset.go:154]: Does this mean the `WasmPath` must begin with `/fluent-bit/wasm`?

Comment (jiuxia211) [reply pkg/operator/daemonset.go:154]: Yes. Because I noticed that's what PositionDB does. Maybe more comments should be added or is there a better way?

Comment (wanjunlei) [reply pkg/operator/daemonset.go:154]: I recommend using [Volumes and VolumesMounts](https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/fluentbit_types.go#L70-L72) to mount wasm files, so you can specify any directory.

Comment (jiuxia211) [reply pkg/operator/daemonset.go]: OK,I have updated it

**Review by wenchajun — APPROVED** [review]

**Review by wanjunlei — APPROVED** [review]

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**